### PR TITLE
Enable token cache for client assertion login flow

### PR DESCRIPTION
**Pull request by isra-fel**
## Description

This PR enables token cache for the client assertion login flow.

- Add client assert code for upcoming change
- cache token for client assertion
- work around AssemblyInformationalVersion issue

## Checklist

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** select appropriate branch. Cmdlets from Autorest.PowerShell should go to [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch. 
* **SHOULD** make the title of PR clear and informative, and in the present imperative tense. 
* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. Add changelog in description section if PR goes into [`generation`](https://github.com/Azure/azure-powershell/tree/generation) branch.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** have approved design review for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) ([_Microsoft internal only_](../blob/main/CONTRIBUTING.md#onboarding)) with following situations
    * Create new module from scratch
    * Create new resource types which are not easy to conform to [Azure PowerShell Design Guidelines](../blob/main/documentation/development-docs/design-guidelines)
    * Create new resource type which name doesn't use module name as prefix
    * Have design question before implementation
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** introduce [breaking changes](../blob/main/documentation/breaking-changes/breaking-changes-definition.md) in Az minor release except preview version.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

File: src/Accounts/Authentication/Identity/Validations.cs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+            {
+                throw new ArgumentException(NoWindowsPowerShellLegacyErrorMessage);
+            }
+
+            return useLegacyPowerShell;
+        }
+
+        private static bool IsValidTenantCharacter(char c)
+        {
+            return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') || (c == '.') || (c == '-');
```
Comment (BethanyZhou) [inline src/Accounts/Authentication/Identity/Validations.cs:90]: return Char.IsLetterOrDigit(c) || (c == '.') || (c == '-');

**Review by BethanyZhou — APPROVED** [review]

**Comment by BethanyZhou** [comment]
Hi @isra-fel, could you check test failure? 

  [xUnit.net 00:00:01.51]     Microsoft.Azure.Commands.Profile.Test.AutosaveTests.EnableAutosaveWhenDisabled [FAIL]
  [xUnit.net 00:00:01.52]     Microsoft.Azure.Commands.Profile.Test.AutosaveTests.EnableAutosaveWhenEnabled [FAIL]
  [xUnit.net 00:00:01.52]     Microsoft.Azure.Commands.Profile.Test.AutosaveTests.DisableAutosaveWhenEnabled [FAIL]
  [xUnit.net 00:00:01.52]     Microsoft.Azure.Commands.Profile.Test.AutosaveTests.DisableAutoSsaveWhenDisabled [FAIL]
    X Microsoft.Azure.Commands.Profile.Test.AutosaveTests.EnableAutosaveWhenDisabled [1ms]
EXEC : error Message:  [D:\a\_work\1\s\build.proj]
     System.InvalidOperationException : The Azure PowerShell session has not been properly initialized.  Please import the module and try again

**Review by dingmeng-xue — APPROVED** [review]

File: src/Accounts/Authentication/Identity/Validations.cs (hunk 1 above)
Comment (isra-fel) [reply src/Accounts/Authentication/Identity/Validations.cs:90]: Good point but the code was originated from azure-sdk repo. See README.md for more info.

**Pull request merged**